crates/sc-proof-of-time/src/pot_state.rs

```diff
@@ -0,0 +1,316 @@
+        Self {
+            config,
+            proof_of_time,
+            state: Mutex::new(InternalState {
+                chain,
+                future_proofs: BTreeMap::new(),
+            }),
+        }
+    }
+
+    /// Extends the chain with the given proof, without verifying it
+    /// (e.g) called when clock maker locally produces a proof.
+    pub fn extend_chain(&self, proof: &PotProof) -> Result<(), PotStateError> {
+        let mut state = self.state.lock();
+        let tip = match state.chain.last() {
+            Some(tip) => tip,
+            None => {
+                self.add_to_tip(&mut state, proof, ExtendAction::MergeFutureProofs);
+                return Ok(());
+            }
+        };
+
+        if (tip.slot_number + 1) == proof.slot_number {
+            self.add_to_tip(&mut state, proof, ExtendAction::MergeFutureProofs);
+            Ok(())
+        } else {
+            // The tip moved by the time the proof was computed.
+            Err(PotStateError::TipMismatch {
+                expected: tip.slot_number + 1,
+                actual: proof.slot_number,
+            })
+        }
+    }
+
+    /// Extends the chain with the given proof, after verifying it
+    /// (e.g) called when the proof is received from a peer via gossip.
+    pub fn verify_and_extend_chain(
+        &self,
+        sender: PeerId,
+        proof: &PotProof,
+    ) -> Result<(), PotStateError> {
+        let mut state = self.state.lock();
+        let tip = match state.chain.last() {
+            Some(tip) => tip.clone(),
+            None => {
+                self.proof_of_time
+                    .verify(proof)
+                    .map_err(PotStateError::InvalidProof)?;
+                self.add_to_tip(&mut state, proof, ExtendAction::MergeFutureProofs);
+                return Ok(());
+            }
+        };
+
+        // Case 1: the proof is for an older slot
+        if proof.slot_number <= tip.slot_number {
+            return Err(PotStateError::StaleProof {
+                tip_slot: tip.slot_number,
+                proof_slot: proof.slot_number,
+            });
+        }
+
+        // Case 2: the proof extends the tip
+        if (tip.slot_number + 1) == proof.slot_number {
+            let expected_seed = tip.next_seed(None);
+            if proof.seed != expected_seed {
+                return Err(PotStateError::InvalidSeed {
+                    expected: expected_seed,
+                    actual: proof.seed,
+                });
+            }
+
+            let expected_key = tip.next_key();
+            if proof.key != expected_key {
+                return Err(PotStateError::InvalidKey {
+                    expected: expected_key,
+                    actual: proof.key,
+                });
+            }
+
+            self.proof_of_time
+                .verify(proof)
+                .map_err(PotStateError::InvalidProof)?;
```

Comment:
> proofs are stateless and are verifiable right away, you just don't know if they correspond to your chain. 
Exactly because verification is costly, I wrote in the spec that we save future proofs until we can see whether they correspond to our chain and only do AES verification if they do and throw away those that do not extend our chain so we save CPU cycles.

Reply:
The maximum value where there can theoretically be PoT forks is the injection depth - two nodes may inject a different value to the seed of PoT creating a fork. Honest PoT inherits forks in consensus.
I’m talking here more about malicious invalid proofs. If we verify AES on future proofs on gossip I can make many of them. If we save them and throw away those that wont extend our local view we potentially save verifying AES on most of them.

Reply:
One of the advantages of doing the verification upfront is that the state management logic (merging future proofs as tip grows) becomes much simpler. Given we don't have injection from consensus yet, we can revisit this later. WDYT?

Some more things we plan to do as discussed with Nazar today:
1. Do the AES verification in gossip validator fn itself
2. Keep track of the validated <key, seed> for a slot number. In the common case of getting the same proof from several honest nodes, this will require AES verification once.
3. For nodes that send proofs that fail validation, reduce reputation 

1, 2 will come immediately, 3 will come later.